heapster-saw/doc/Permissions.md

```diff
@@ -0,0 +1,123 @@
+| `llvmptr w`           | llvm-specific crucible type of pointers of width `w` |
+| `rwmodality`          | type of modality for read or write permissions |
+| `permlist`            | type of a list of permissions - outdated |
+| `llvmframe w`         | type of ghost objects that represent the current stack frame with respect to bitvectors of width `w` |
+| `perm(a)`             | type of permissions describing how to use an object of type `a`. That is, the proposition `x:p` for a permission `p` of type `perm(a)` means "`a` value `x` of type `a` has permission `p`". |
+| `llvmshape w`         | type of shapes with respect to bitvectors of width `w` |
+| `llvmblock w`         | type of blocks of memory with respect to bitvectors of width `w` |
+
+Permission expressions
+======================
+
+Expressions that are considered \"pure\" for use in permissions.
+
+A llvm-pointer (aka an llvm value) is either an llvmword or a variable+an offset
+
+Any crucible type can have a variable of that type, and that thing is an expression
+
+| Permission Expressions `e` | Type | Description |
+| :---: | :---: | :--- |
+| `x`                          | any                | Any expression variable |
+| `unit`                       | `unit`             | A unit literal |
+| `b`                          | `bool`             | A literal boolean value |
+| `n`                          | `nat`              | A literal natural number |
+| `n`                          | `bv w`             | A literal bitvector |
+| `b1 + b2`                    | `bv w`             | Sum of two bitvectors |
+| `b1 * b2`                    | `bv w`             | Linear multiplication of two bitvectors, meaning that one of the operands must be a constant |
+| `struct(e1,..,en)`           | `struct(a1,..,an)` | A (crucible) struct is a tuple of expressions for each argument of the struct type. Crucible structs are different from C structs, and we only use crucible structs when we need to, otherwise C structs are described manually as pointers into chunks of memory |
+| `llvmword(e)`                | `llvmptr(w)`       | An LLVM value that represents a word, i.e. whose region identifier is 0, given a bitvector expression `e:bv w` |
+| `x &+ o`                     | `llvmptr(w)`       | An LLVM pointer built by adding an offset `o` (a bitvector expression) to an LLVM pointer variable `x` |
+| `R`                          | `rwmodality`       | The read modality |
+| `W`                          | `rwmodality`       | The write modality |
+| `p`                          | `perm(a)`          | A permission as an expression |
+
+In addition to the above expressions, we also have shape expressions, which we separated out only for the sake of readability.
+
+| Shape Permission Expressions `sh` | Type | Description |
+| :---: | :---: | :--- |
+| `emptysh`                   | `llvmshape w` | The empty shape |
```

Comment:
```suggestion
| `emptysh`                   | <nobr>`llvmshape w`</nobr> | The empty shape |
```
On my machine this column is formatted with awkward lines breaks, I believe this change should fix that for the entire column.

Reply:
This didn't work for some reason so it should probably be reverted - sorry...

Reply:
Unfortunately this didn't work for me on chrome on linux. I've tried a few other formatting tricks and haven't found a good solution yet. Keeping it in if it improves formatting for some machines.

Reply:
Ok, I'll revert this. If anyone finds a way to make it prettier in the future let's do it, but I think it's legible enough for now. :)